File: ios/VineTrackV2/App/PinDuplicateChecker.swift
import Foundation
import CoreLocation

/// Pure helpers for warning when a new pin is being dropped on top of an
/// existing one. Backend-neutral — works with whatever pins are already
/// loaded into `MigratedDataStore.pins`.
nonisolated enum PinDuplicateChecker {

    /// Default fallback radius (metres) when no row-spacing is known.
    /// Sized for typical field GPS error (3-5 m) so two pins placed at the
    /// same vine actually trigger a duplicate warning even when the GPS
    /// fix has drifted between drops.
    static let fallbackRadiusMeters: Double = 3.0

    /// Hard cap on radius. Above this, two pins are clearly different
    /// targets — but this is intentionally generous to account for GPS
    /// noise during active trips at slow speeds.
    static let maxRadiusMeters: Double = 6.0

    /// Minimum radius even when row spacing is known. Bumped from 1.5 m to
    /// 2.5 m so vineyard GPS jitter (typically 2–3 m horizontal accuracy)
    /// can't sneak a near-duplicate pin past the warning. Narrow-row
    /// paddocks may now warn on adjacent rows, which is the correct
    /// trade-off for repair/growth pin work.
    static let minRadiusMeters: Double = 2.5

    /// Compute the duplicate-warning radius for a pin being dropped at
    /// `coordinate`. Uses half the row spacing of the most relevant paddock
    /// (the one containing the coordinate, falling back to `paddockId`),
    /// or a conservative fallback when geometry isn't available.
    static func duplicateRadius(
        coordinate: CLLocationCoordinate2D,
        paddockId: UUID?,
        paddocks: [Paddock]
    ) -> Double {
        if let containing = RowGuidance.paddock(for: coordinate, in: paddocks),
           containing.rowWidth > 0 {
            return min(maxRadiusMeters, max(minRadiusMeters, containing.rowWidth / 2.0))
        }
        if let id = paddockId,
           let paddock = paddocks.first(where: { $0.id == id }),
           paddock.rowWidth > 0 {
            return min(maxRadiusMeters, max(minRadiusMeters, paddock.rowWidth / 2.0))
        }
        return fallbackRadiusMeters
    }

    /// The closest pin within `radius` of `coordinate`, scoped to the same
    /// vineyard. Active (not-completed) pins are preferred; completed pins
    /// are returned only when no active match exists. Returns `nil` when
    /// no pin is in range.
    /// Along-row duplicate radius. When pin-snapping projects two pins
    /// onto the same row line, raw GPS distance under-counts duplicates
    /// because tractor jitter spreads samples along the row. We use a
    /// tighter, fixed along-row radius for same-row, same-mode pins so a
    /// repair tapped twice within a couple of metres warns reliably.
    static let alongRowDuplicateMetres: Double = 2.5

    /// Find a likely duplicate using along-row geometry. Same vineyard +
    /// same paddock + same row number + same mode + along-row distance
    /// within `alongRowDuplicateMetres`. Returns the closest match (open
    /// pins preferred). Falls back to nil when no row context exists —
    /// callers should then use the lat/lng-based `nearbyPin` check.
    static func nearbyPinAlongRow(
        snappedCoordinate: CLLocationCoordinate2D,
        vineyardId: UUID?,
        paddockId: UUID?,
        rowNumber: Int?,
        side: PinSide? = nil,
        mode: PinMode?,
        in pins: [VinePin],
        paddocks: [Paddock]
    ) -> (pin: VinePin, distance: Double)? {
        guard let paddockId, let rowNumber,
              let paddock = paddocks.first(where: { $0.id == paddockId }) else {
            return nil
        }
        guard let snappedSelf = RowGuidance.snapToRow(
            coordinate: snappedCoordinate,
            rowNumber: rowNumber,
            in: paddock
        ) else { return nil }

        var bestActive: (pin: VinePin, distance: Double)?
        var bestDone: (pin: VinePin, distance: Double)?
        for pin in pins {
            if let vid = vineyardId, pin.vineyardId != vid { continue }
            guard pin.paddockId == paddockId else { continue }
            // Match by the new pin_row_number when present (actual vine
            // row), otherwise fall back to the legacy row_number storage
            // (driving path floor). New + legacy pins both compare cleanly.
            let candidateRow = pin.pinRowNumber ?? pin.rowNumber
            guard candidateRow == rowNumber else { continue }
            // Only constrain by side when the caller actually knows the
            // side — otherwise treat both sides as candidate duplicates.
            // Prefer the new pin_side; legacy pin.side is operator-side too.
            if let side, (pin.pinSide ?? pin.side) != side { continue }
            if let mode, pin.mode != mode { continue }
            guard let snappedOther = RowGuidance.snapToRow(
                coordinate: pin.attachedCoordinate,
                rowNumber: rowNumber,
                in: paddock
            ) else { continue }
            let delta = abs(snappedSelf.distanceAlongMetres - snappedOther.distanceAlongMetres)
            guard delta <= alongRowDuplicateMetres else { continue }
            let scoped: (VinePin, Double) = (pin, delta)
            if pin.isCompleted {
                if bestDone == nil || delta < bestDone!.distance { bestDone = scoped }
            } else {
                if bestActive == nil || delta < bestActive!.distance { bestActive = scoped }
            }
        }
        return bestActive ?? bestDone
    }

    static func nearbyPin(
        coordinate: CLLocationCoordinate2D,
        vineyardId: UUID?,
        paddockId: UUID?,
        radius: Double,
        in pins: [VinePin]
    ) -> (pin: VinePin, distance: Double)? {
        var bestActive: (pin: VinePin, distance: Double)?
        var bestDone: (pin: VinePin, distance: Double)?

        for pin in pins {
            if let vid = vineyardId, pin.vineyardId != vid { continue }
            let d = RowGuidance.metresBetween(
                coordinate,
                CLLocationCoordinate2D(latitude: pin.latitude, longitude: pin.longitude)
            )
            guard d <= radius else { continue }

            // Same paddock matches sort first by tightening the radius slightly.
            let scoped: (VinePin, Double) = (pin, d)
            if pin.isCompleted {
                if bestDone == nil || d < bestDone!.distance {
                    bestDone = scoped
                }
            } else {
                if bestActive == nil || d < bestActive!.distance {
                    bestActive = scoped
                }
            }
            _ = paddockId
        }
        return bestActive ?? bestDone
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/PinDuplicateChecker.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.model.Paddock
import com.rork.vinetrack.data.model.Pin
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sqrt

/**
 * Pure, backend-neutral duplicate detector for Repairs/Growth launcher pins.
 *
 * Two complementary paths, both mirroring the iOS `PinDuplicateChecker` flow
 * (`RepairsGrowthView.checkDuplicate` runs along-row first, raw distance second):
 *
 * 1. [nearbyAlongRow] — preferred. When a new launcher pin has been snapped to a
 *    vine row, warn if there is already an *open* pin on the same block + same
 *    attached row + same side + same mode whose stored along-row position is
 *    within [ALONG_ROW_DUPLICATE_METRES]. Catches a repair tapped twice on the
 *    same vine even when GPS jitter spreads the raw fixes apart.
 *
 * 2. [nearbyRawDistance] — legacy fallback. Older pins were created before
 *    row-attachment existed, so they carry no `pin_row_number` /
 *    `along_row_distance_m` and never participate in the along-row path. This
 *    fallback still catches them using raw GPS distance, scoped conservatively
 *    to the same block / mode / side / manual row so it stays safe.
 *
 * Matching notes (parity with iOS):
 * - iOS re-snaps each existing pin's raw coordinate to the row at check time;
 *   Android compares the **stored** `along_row_distance_m` in the along-row path.
 * - iOS' raw fallback (`nearbyPin`) scopes only by vineyard + radius. Android's
 *   fallback is intentionally stricter (block + mode + side + manual row) to
 *   avoid false positives across blocks, while keeping the same threshold.
 * - Matching is by mode (Repairs vs Growth), not by category — same as iOS.
 * - Soft-deleted pins are never matched; open pins are preferred.
 */
object PinDuplicateChecker {

    /** Along-row radius (m) within which two same-row pins are treated as duplicates. */
    const val ALONG_ROW_DUPLICATE_METRES: Double = 2.5

    /**
     * Raw-distance fallback radii, mirroring iOS `PinDuplicateChecker`:
     * half the block's row spacing, clamped to [MIN_RADIUS_METRES, MAX_RADIUS_METRES],
     * or [FALLBACK_RADIUS_METRES] when the block has no known row width.
     */
    const val FALLBACK_RADIUS_METRES: Double = 3.0
    const val MIN_RADIUS_METRES: Double = 2.5
    const val MAX_RADIUS_METRES: Double = 6.0

    private const val METRES_PER_DEG_LAT = 111_320.0

    /**
     * A likely-duplicate match. [alongRow] distinguishes the two paths so the UI
     * can word the warning correctly: along-row matches know the attached row and
     * exact along-row distance; raw-distance fallbacks only know straight-line
     * GPS distance within the block.
     */
    data class Match(
        val pin: Pin,
        val distanceM: Double,
        val alongRow: Boolean,
    )

    /**
     * Find the nearest likely duplicate of a pin whose resolved [candidate]
     * attachment sits on a known vine row inside [paddockId]. Returns null when
     * there is no row context or no open same-row/side/mode pin within range.
     */
    fun nearbyAlongRow(
        candidate: RowAttachment.Attachment,
        paddockId: String?,
        mode: String?,
        pins: List<Pin>,
    ): Match? {
        if (paddockId == null) return null
        val candidateSide = candidate.pinSide?.lowercase()?.takeIf { it == "left" || it == "right" }
        val candidateMode = mode?.lowercase()?.takeIf { it.isNotBlank() }

        var best: Match? = null
        for (pin in pins) {
            if (pin.deletedAt != null) continue
            if (pin.isCompleted) continue
            if (pin.paddockId != paddockId) continue

            // Same attached vine row (snapped row), tolerant to fractional values.
            val existingRow = pin.pinRowNumber ?: continue
            if (existingRow != candidate.pinRowNumber) continue

            // Constrain by side only when both sides are known.
            if (candidateSide != null) {
                val existingSide = (pin.pinSide ?: pin.side)?.lowercase()
                if (existingSide == "left" || existingSide == "right") {
                    if (existingSide != candidateSide) continue
                }
            }

            // Constrain by mode (Repairs vs Growth) when known.
            if (candidateMode != null) {
                val existingMode = pin.mode?.lowercase()
                if (!existingMode.isNullOrBlank() && existingMode != candidateMode) continue
            }

            val existingAlong = pin.alongRowDistanceM ?: continue
            val delta = abs(existingAlong - candidate.alongRowDistanceM)
            if (delta > ALONG_ROW_DUPLICATE_METRES) continue

            if (best == null || delta < best.distanceM) {
                best = Match(pin = pin, distanceM = delta, alongRow = true)
            }
        }
        return best
    }

    /**
     * Legacy fallback for older pins lacking row attachment. Uses raw GPS
     * distance from the new pin's [latitude]/[longitude] to each existing pin's
     * stored coordinate, within a row-spacing-derived radius.
     *
     * Only pins that did NOT participate in [nearbyAlongRow] are considered here
     * (i.e. those missing `pin_row_number` or `along_row_distance_m`), so the two
     * paths never double-warn for the same pin. Matching is scoped to the same
     * block, same mode (when known), same side (when both present) and same
     * manual `row_number` (when both present) to keep it conservative.
     */
    fun nearbyRawDistance(
        latitude: Double?,
        longitude: Double?,
        paddockId: String?,
        mode: String?,
        side: String?,
        manualRowNumber: Int?,
        paddock: Paddock?,
        pins: List<Pin>,
    ): Match? {
        if (latitude == null || longitude == null) return null
        if (paddockId == null) return null
        val candidateSide = side?.lowercase()?.takeIf { it == "left" || it == "right" }
        val candidateMode = mode?.lowercase()?.takeIf { it.isNotBlank() }
        val radius = duplicateRadius(paddock)

        var best: Match? = null
        for (pin in pins) {
            if (pin.deletedAt != null) continue
            if (pin.isCompleted) continue
            if (pin.paddockId != paddockId) continue

            // Only legacy pins missing row attachment — row-attached pins are
            // handled by the (preferred) along-row path.
            if (pin.pinRowNumber != null && pin.alongRowDistanceM != null) continue

            val existingLat = pin.latitude ?: continue
            val existingLng = pin.longitude ?: continue

            // Constrain by side only when both sides are known.
            if (candidateSide != null) {
                val existingSide = (pin.pinSide ?: pin.side)?.lowercase()
                if (existingSide == "left" || existingSide == "right") {
                    if (existingSide != candidateSide) continue
                }
            }

            // Constrain by mode (Repairs vs Growth) when known.
            if (candidateMode != null) {
                val existingMode = pin.mode?.lowercase()
                if (!existingMode.isNullOrBlank() && existingMode != candidateMode) continue
            }

            // Constrain by manual row number only when both pins have one.
            if (manualRowNumber != null && pin.rowNumber != null) {
                if (pin.rowNumber != manualRowNumber) continue
            }

            val d = metresBetween(latitude, longitude, existingLat, existingLng)
            if (d > radius) continue

            if (best == null || d < best.distanceM) {
                best = Match(pin = pin, distanceM = d, alongRow = false)
            }
        }
        return best
    }

    /**
     * Raw-distance warning radius: half the block's row spacing clamped to
     * [MIN_RADIUS_METRES, MAX_RADIUS_METRES], falling back to
     * [FALLBACK_RADIUS_METRES]. Mirrors iOS `PinDuplicateChecker.duplicateRadius`.
     */
    fun duplicateRadius(paddock: Paddock?): Double {
        val width = paddock?.rowWidth
        return if (width != null && width > 0) {
            (width / 2.0).coerceIn(MIN_RADIUS_METRES, MAX_RADIUS_METRES)
        } else {
            FALLBACK_RADIUS_METRES
        }
    }

    /** Straight-line distance (m) between two lat/lng points (equirectangular). */
    private fun metresBetween(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double,
    ): Double {
        val midLat = (lat1 + lat2) / 2.0
        val mPerDegLon = METRES_PER_DEG_LAT * cos(midLat * Math.PI / 180.0)
        val dx = (lng2 - lng1) * mPerDegLon
        val dy = (lat2 - lat1) * METRES_PER_DEG_LAT
        return sqrt(dx * dx + dy * dy)
    }
}
